zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/LearnerSyncThrottlerTest.java

```diff
@@ -0,0 +1,169 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import org.apache.zookeeper.ZKTestCase;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@RunWith(Parameterized.class)
+public class LearnerSyncThrottlerTest extends ZKTestCase {
+    private static final Logger LOG = LoggerFactory.getLogger(LearnerSyncThrottlerTest.class);
+
+    private LearnerSyncThrottler.SyncType syncType;
+    public LearnerSyncThrottlerTest(LearnerSyncThrottler.SyncType syncType){
+        this.syncType = syncType;
+    }
+
+    @Parameterized.Parameters
+    public static Collection syncTypes() {
+        return Arrays.asList(new Object[][]{
+                {LearnerSyncThrottler.SyncType.DIFF}, {LearnerSyncThrottler.SyncType.SNAP}});
+    }
+    @Test(expected = SyncThrottleException.class)
+    public void testTooManySyncsNonessential() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(5, syncType);
+        for (int i = 0; i < 6; i++) {
+            throttler.beginSync(false);
+        }
+    }
+
+    @Test(expected = SyncThrottleException.class)
+    public void testTooManySyncsEssential() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(5, syncType);
+        try {
+            for (int i = 0; i < 6; i++) {
+                throttler.beginSync(true);
+            }
+        }
+        catch (SyncThrottleException ex) {
+            Assert.fail("essential syncs should not be throttled");
+        }
+        throttler.endSync();
+        throttler.beginSync(false);
+    }
+
+    @Test
+    public void testNoThrottle() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(5, syncType);
+        try {
+            for (int i = 0; i < 6; i++) {
+                throttler.beginSync(true);
+            }
+        }
+        catch (SyncThrottleException ex) {
+            Assert.fail("essential syncs should not be throttled");
+        }
+        throttler.endSync();
+        for (int i = 0; i < 5; i++) {
+            throttler.endSync();
+            throttler.beginSync(false);
+        }
+        Assert.assertTrue("should get here without exception", true);
+    }
+
+    @Test
+    public void testTryWithResourceNoThrottle() throws Exception {
+        LearnerSyncThrottler throttler =
+                new LearnerSyncThrottler(1, syncType);
+        for (int i = 0; i < 3; i++) {
+            throttler.beginSync(false);
+            try {
+                Assert.assertEquals(1, throttler.getSyncInProgress());
+            } finally {
+                throttler.endSync();
+            }
+        }
+    }
+
+    @Test(expected = SyncThrottleException.class)
+    public void testTryWithResourceThrottle() throws Exception {
```

Comment:
It is not very clear that we are testing that the second call to beginSync is the one that throws the exception
I suggest to use an explicit catch block and not use the 'expected' attribute

Reply:
done